### PR TITLE
Added Cyton Status check

### PR DESCRIPTION
**Pull request by MouradLachhab**
Added a request for Cyton system status to make sure it is available when connecting.

Note: I have noticed that when closing by usb_serial port, I get a java.lang.InterruptedException. I have tried to catch that error or find how to avoid it without success, so if you have any idea.

image

### CODE REVIEW

**Review by abelfodil — DISMISSED** [review]

**Comment by abelfodil** [comment]
The exception seems to come from here:

https://github.com/felHR85/UsbSerial/blob/master/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L72

https://github.com/felHR85/UsbSerial/blob/319cb9de686325dcc2814d7cbb6ed7dd36011183/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L121

**Comment by MouradLachhab** [comment]
> The exception seems to come from here:
> 
> https://github.com/felHR85/UsbSerial/blob/master/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L72
> 
> https://github.com/felHR85/UsbSerial/blob/319cb9de686325dcc2814d7cbb6ed7dd36011183/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L121

**Comment by MouradLachhab** [comment]
> The exception seems to come from here:
> 
> https://github.com/felHR85/UsbSerial/blob/master/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L72
> 
> https://github.com/felHR85/UsbSerial/blob/319cb9de686325dcc2814d7cbb6ed7dd36011183/usbserial/src/main/java/com/felhr/usbserial/SerialBuffer.java#L121

Yea, I've looked into it, and the latest commit for that file was exactly to fix an issue with the thread not stopping when the port was closing: 
https://github.com/felHR85/UsbSerial/commit/001e745a4aa3e552ac82f6823316515a0a1393c7

But I'm still not sure how I'm supposed to deal with this. As of right now, simply opening the port and closing it right after will trigger this exception.

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 1]
```diff
@@ -46,11 +47,39 @@ class SerialRepository implements IAcquisitionDeviceRepository {
...
+    String status = "";
+
+    _inputStreamSubscription = _serialPort.inputStream.listen((event) {
+      status += String.fromCharCodes(event);
+
+      if (isFullMessage(status)) {
+        _inputStreamSubscription.cancel();
+
+        if (status == CYTON_SYSTEM_UP)
+          callback(true, null);
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:63]: Use optional parameter for the Exception using [].

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 2]
```diff
@@ -46,11 +47,39 @@ class SerialRepository implements IAcquisitionDeviceRepository {
     _serialPort.setPortParameters(
         115200, UsbPort.DATABITS_8, UsbPort.STOPBITS_1, UsbPort.PARITY_NONE);
 
-    callback(true, null);
+    checkCytonConnection(callback);
+  }
+
+  Future<void> checkCytonConnection(Function(bool, Exception) callback) async {
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:53]: This methode is private. Use _ as a prefix for its name.

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 3]
```diff
@@ -46,11 +47,39 @@ class SerialRepository implements IAcquisitionDeviceRepository {
...
+
+  bool isFullMessage(String message) {
+    return message.length > CYTON_MESSAGE_FOOTER.length &&
+        message.substring(message.length - CYTON_MESSAGE_FOOTER.length) ==
+            CYTON_MESSAGE_FOOTER;
   }
 
   Future<void> disconnect() async {
     await _serialPort?.close();
+    _inputStreamSubscription = null;
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:82]:
```suggestion
    _inputStreamSubscription?.cancel();
```

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 4]
```diff
@@ -46,11 +47,39 @@ class SerialRepository implements IAcquisitionDeviceRepository {
     _serialPort.setPortParameters(
         115200, UsbPort.DATABITS_8, UsbPort.STOPBITS_1, UsbPort.PARITY_NONE);
 
-    callback(true, null);
+    checkCytonConnection(callback);
+  }
+
+  Future<void> checkCytonConnection(Function(bool, Exception) callback) async {
+    String status = "";
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:54]: Le linter ne dit pas que le type est implicite ici? On devrait pourvoir simplement utiliser `var status = "";`

**Review by WilliamHarvey97 — APPROVED** [review]

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart (hunk 4 above)
Comment (MouradLachhab) [reply mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:55]: Nope, je sais pas si c'est une setting du linter que je dois activé or something.

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 5]
```diff
@@ -12,6 +12,7 @@ class SerialRepository implements IAcquisitionDeviceRepository {
   UsbPort _serialPort;
   List<AcquisitionDevice> _acquisitionDevicePersistency = [];
   List<UsbDevice> _serialDevices = [];
+  StreamSubscription _inputStreamSubscription;
   final streamController = StreamController<List<AcquisitionDevice>>();
 
   void initializeRepository() {
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:18]: @override

File: mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart [hunk 6]
```diff
@@ -46,11 +47,40 @@ class SerialRepository implements IAcquisitionDeviceRepository {
...
+    await _serialPort.write(Uint8List.fromList(CYTON_GET_STATUS));
+  }
+
+  bool isFullMessage(String message) {
+    return message.length > CYTON_MESSAGE_FOOTER.length &&
+        message.substring(message.length - CYTON_MESSAGE_FOOTER.length) ==
+            CYTON_MESSAGE_FOOTER;
   }
 
   Future<void> disconnect() async {
```
Comment (WilliamHarvey97) [inline mobile/lib/src/infrastructure/connection_repositories/serial_repository.dart:81]: @override

**Pull request merged**